software/firmware/europi_script.py

```diff
@@ -52,23 +52,111 @@ def display_name(cls):
         """
         return cls.__qualname__
 
-    #
-    # private functions for now, with the intention that they are eventually added to the public API
-    #
+    # Save State Methods
+
+    @property
+    def _state_filename(self):
+        return f"saved_state_{self.__class__.__qualname__}.txt"
+
+    def save_state(self, state: str):
+        """Take state in persistence format as a string and write to disk."""
+        return self._save_state(state)
+
+    def save_state_bytes(self, state: bytes):
+        """Take state in persistence format as bytes and write to disk.
+
+        .. note::
+            Be mindful of how often `save_state_bytes()` is called because
+            writing to disk too often can slow down the performance of your
+            script. Only call save state when state has changed and consider
+            adding a time since last save check to reduce save frequency.
+
+        ::
+
+            # https://docs.python.org/3/library/struct.html#format-characters
+            format_string = "b?"
+            state = pack(format_string, 2, True)
+            super().save_state_json(state)
+        """
+        return self._save_state(state, mode='wb')
+
+    def save_state_json(self, state: dict):
+        """Take state as a dict and save as a json string.
+
+        .. note::
+            Be mindful of how often `save_state_json()` is called because
+            writing to disk too often can slow down the performance of your
+            script. Only call save state when state has changed and consider
+            adding a time since last save check to reduce save frequency.
 
-    def _save_state(self, state: str):
-        with open(self._state_filename, 'w') as file:
+        ::
+
+            # Save the current state variables as JSON.
+            state = {
+                "counter": self.counter,  # int
+                "enabled": self.enabled,  # bool
+            }
+            super().save_state_json(state)
+        """
+        json_str = json.dumps(state)
+        return self._save_state(json_str)
+
+    def _save_state(self, state: str, mode: str ='w'):
+        with open(self._state_filename, mode) as file:
             file.write(state)
 
-    def _load_state(self) -> str:
+    def load_state(self) -> str:
+        """Check disk for saved state, if it exists, return the raw state value as a string."""
+        return self._load_state()
+
+    def load_state_bytes(self) -> bytes:
+        """Check disk for saved state, if it exists, return the raw state value as bytes.
+        
+        ::
+
+            # Check for a previously saved state. If it exists, return state as a
+            # byte string. If no state is found, an empty string will be returned.
```

Comment:
> Check for a previously saved state. If it exists, return state as a byte string.

This is redundant with the docstring.

>  If no state is found, an empty string will be returned.

This should be part of the docstring, not a comment in the example.